Bug 1455432: Audit log plugin should allow to include or exclude

gl-sergei opened the pull request:
logging for specific users

Added two global variables:
- audit_log_include_accounts: accounts to include in audit logging.
- audit_log_exclude_accounts: accounts to exclude from audit logging.

The value can be NULL, empty string or comma separated list of
accounts in form user@host or 'user'@'host' (if user or host contains
comma).

Only one variable can be not empty at a time. If one is set to be not
empty, another one becomes empty.

gl-sergei commented:
Feature was created to be similar to https://dev.mysql.com/doc/refman/5.7/en/audit-log-legacy-filtering.html

gl-sergei commented:
Similar feature in MariaDB: https://mariadb.com/kb/en/mariadb/about-the-mariadb-audit-plugin/#excluding-or-including-users

gl-sergei commented:
New https://dev.mysql.com/doc/refman/5.7/en/audit-log-filtering.html is possible to implement only in 5.7. Eventually, we may go to that direction.

gl-sergei commented:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1188/

laurynas-biveinis commented:
This needs blueprints too (it looks like this implements a subset of https://blueprints.launchpad.net/percona-server/+spec/audit-log-filtering)

laurynas-biveinis commented on mysql-test/t/audit_log_filter_users.test:
```diff
@@ -0,0 +1,83 @@
+  print "================================\n";
+EOF
+--remove_file $MYSQLD_DATADIR/test_audit.log
+--remove_file $MYSQLD_DATADIR/test_audit_csv.log
```
The testcase only ever connects with "root",  thus it cannot possibly cover the functionality, please create some users, and merge the "parser" (lines 4-24) tests with functionality tests, using different users

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -775,6 +787,32 @@ int is_event_class_allowed_by_policy(unsigned int class,
+  {
+    const struct mysql_event_connection *event_connection=
+      (const struct mysql_event_connection *) event;
+    local->skip_logging= 0;
```
s/0/FALSE

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -775,6 +787,32 @@ int is_event_class_allowed_by_policy(unsigned int class,
+                                          event_connection->user_length,
+                                          event_connection->host,
+                                          event_connection->host_length))
+      local->skip_logging= 1;
```
s/1/TRUE, also I'd add a DBUG_ASSERT(audit_log_exclude_accounts == NULL), maybe even an early return

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -775,6 +787,32 @@ int is_event_class_allowed_by_policy(unsigned int class,
+                                         event_connection->user_length,
+                                         event_connection->host,
+                                         event_connection->host_length))
+      local->skip_logging= 1;
```
s/1/TRUE, DBUG_ASSERT(audit_log_include_accounts == NULL)

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -798,11 +839,14 @@ void audit_log_notify(MYSQL_THD thd __attribute__((unused)),
           strncmp(event_general->general_command, "Quit", 4) == 0)
         break;
 
+      if (local->skip_logging)
```
Move this up right after the audit_log_update_thd_local call?

gl-sergei replied:
It depends on what we want to do actually. If we want to skip connect/disconnect/change user events, then yes. If we don't want then we can only move it before `strncmp(event_general->general_command.str, "Quit", 4)`

laurynas-biveinis replied:
Probably we would want to skip connect/disconnect events for filtered users but not change user events to/from unfiltered/filtered to filtered/unfiltered (this would be a good addition to the testcase too)

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+
+static PSI_rwlock_key key_LOCK_account_list;
+static PSI_rwlock_info all_rwlock_list[]=
+{{ &key_LOCK_account_list, "audit_log_filter::account_list", PSI_FLAG_GLOBAL}};
```
Is the "::" convention used somewhere else for key names like that? I thought it was "_" everywhere

gl-sergei replied:
Yup. In `buffer.c`:

```
static PSI_mutex_info mutex_key_list[]=
{{ &key_log_mutex, "audit_log_buffer::mutex", PSI_FLAG_GLOBAL}};
```

But, again, I don't mind to change convention to whatever is common in server code.

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+                        const char *host, size_t host_length)
+{
+  account *acc= (account *) my_malloc(sizeof(account),
+                                      MYF(MY_FAE | MY_ZEROFILL));
```
No need for MY_ZEROFILL? account_init will write everything anyway

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+
+  account_init(acc, user, user_length, host, host_length);
+
+  DBUG_ASSERT(user_length + host_length + 2 <= sizeof(acc->name));
```
copy paste...

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+  memcpy(acc->name, user, user_length);
+  memcpy(acc->name + user_length + 1, host, host_length);
+  acc->name[user_length]= '@';
+  acc->name[user_length + host_length + 1]= 0;
```
...leftover?

gl-sergei replied:
Yup. account_init was made of account_create and I just forgot to remove the rest...

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+  Get account key
+*/
+uchar* account_get_key(const account *acc, size_t *length,
+                       my_bool not_used __attribute__((unused)))
```
Upstream just cleaned their codebase with `s/__attribute__/MY_ATTRIBUTE`, we should probably do the same

gl-sergei replied:
This was actually copied from somewhere in server! and simple grep for `my_bool not_used __attribute_` gives 40 matches :)

gl-sergei replied:
OK. Probably it is not in my GCA yet. `MY_ATTRIBUTE` isn't even defined anywhere.

laurynas-biveinis replied:
Yes - the upstream cleanup is only in 5.6.31. So here `__attribute__` is OK then

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+  Remove enclosing quotes from string if any.
+*/
+static
+void unqote_string(char *string, size_t *string_length)
```
s/unqote/unquote

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+  if (string[0] == '\'' && string[*string_length - 1] == '\'')
+  {
+    memmove(string, string + 1, *string_length);
+    *string_length= *string_length - 2;
```
Can actually move two bytes less by moving this above memmove

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+  char *string_copy= my_strdup(string, MYF(MY_FAE));
+  char *entry= string_copy;
+  int string_length= strlen(string_copy);
+  my_bool quote;
```
declare in the loop

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+static
+void account_list_from_string(HASH *hash, const char *string)
+{
+  size_t entry_length;
```
declare in the loop

laurynas-biveinis commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,242 @@
+void audit_log_filter_init()
+{
+#ifdef HAVE_PSI_INTERFACE
+  mysql_rwlock_register("server_audit", all_rwlock_list,
```
PFS key name should share the prefix with other keys in the plugin

gl-sergei replied:
Do you mean category? There is a problem here already.

``` c++
#ifdef HAVE_PSI_INTERFACE
  if(PSI_server)
  {
    PSI_server->register_mutex("server_audit",
                               mutex_key_list, array_elements(mutex_key_list));
    PSI_server->register_cond("server_audit",
                              cond_key_list, array_elements(cond_key_list));
  }
#endif /* HAVE_PSI_INTERFACE */
```

and in `logger.h` (weird, no?):

```
#define PSI_CATEGORY "audit_logger"
```

and then in `audit_log.c`

```
  mysql_memory_register(PSI_CATEGORY, all_audit_log_memory, count);
```

But thats in 5.7.

In 5.6 we have `"server_audit"` in `buffer.c` and `"audit_logger"` in `file_logger.c` (which is taken from MariaDB).

I can address it in this PR and make sure it consistent in 5.7 as well.

laurynas-biveinis replied:
Thanks - please do

laurynas-biveinis commented:
IMHO this should be more strict with include/exclude vars both being non-NULL at the same time: fail setting NULL->non-NULL if the other is non-NULL. This way user is forced to reset the other first, confirming that he knows what he's doing

gl-sergei commented:
I agree that automatically change `exclude` when user set `include` is bad. I will implement CHECK functions for these variables.

gl-sergei commented:
changes:
- behaviour change: do not allow to change `include_accounts` when
   setting `exclude_accounts` and vice versa.
- `TRUE/FALSE` instead of `1/0` for `skip_logging`
- removed leftover redundant code in `account_create`
- `memmove` and `*string_length-= 2` swapped
- fixed typo in `unquote_string`
- `account_list_from_string`: `entry_length` and `quote` declarations
   moved inside the `while` loop.
- use AUDIT_LOG_PSI_CATEGORY everywhere when registering PSI objects
- always log "Change user" events, allow to filter out
   Connect/Disconnect/Query events.

gl-sergei commented:
Blueprint: https://blueprints.launchpad.net/percona-server/+spec/audit-filter-users
Jenkins: http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1202/

gl-sergei commented:
Note, that accounts in `exclude_accounts` and `include_accounts` are not the ones from `mysql.user` table. They are as logged in audit log file (fields user and host). It is possible to use `security_context_service` in `5.7` to get `priv_user` and `priv_host`, but I don't know how to achieve same behaviour in `5.6`

gl-sergei commented:
As an option we can backport `security_context_service` to 5.6 and check in plugin it's presence (for the case when user decided to use new version of plugin with older server). If it is not available, we can fallback to current behaviour.

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -775,6 +787,32 @@ int is_event_class_allowed_by_policy(unsigned int class,
+                                          event_connection->user_length,
+                                          event_connection->host,
+                                          event_connection->host_length))
+      local->skip_logging= TRUE;
```
Previous review round comments here?

gl-sergei replied:
I added assert in `audit_log_include_accounts_update` and `audit_log_exclude_accounts_update`

laurynas-biveinis replied:
asserts here won't hurt here neither -  this function also needs only one of the two variables != NULL in order to work correctly

gl-sergei replied:
Added assert at the beginning of the function, it is more strict than place it here.

laurynas-biveinis commented:
The current implementation logs change_user from A to B, even if, say, both A and B are excluded?

laurynas-biveinis commented:
No need to backport security_context_service IMHO

gl-sergei commented:
Yes, change_user will be logged regardless of filtering. I think it is good enough behaviour.

gl-sergei commented:
OK, then I think we won't use `security_context_service` in 5.7 to make no confusion.

laurynas-biveinis commented:
I think it's OK not to use it in 5.6 and use it in 5.7, and document any user-visible differences in "what's changed in 5.7" doc, because that's what has been done for other features. But what are the actual user-facing differences?

gl-sergei commented:
Changes are rather big. Lets say you have account 'user1'@'%', and you connected via TCP from localhost. Then, your user/host will be 'user1'@'localhost', but priv_user/priv_host (actual authenticated account) will be 'user1'@'%'. Pair user/host also may change when using proxy users and so on.

gl-sergei commented:
What else is needed for this PR? If this one is OK, I'll update 5.7 version.

laurynas-biveinis commented:
I am not sure why change_user is exempt for filtering. Except for that, 5.6 LGTM

gl-sergei commented:
I checked Oracle's audit plugin and it doesn't exempt `Change user` from
filtering. It also doesn't log `Change user` `STATUS` event. Patch
changed to match oracle's plugin behavior.

laurynas-biveinis commented:
LGTM

gl-sergei commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,234 @@
+
+typedef struct
+{
+  char name[USERNAME_LENGTH + HOSTNAME_LENGTH];
```
Should be `USERNAME_LENGTH + HOSTNAME_LENGTH + 1`

gl-sergei commented on plugin/audit_log/filter.c:
```diff
@@ -0,0 +1,234 @@
+  char *string_copy= my_strdup(string, MYF(MY_FAE));
+  char *entry= string_copy;
+  int string_length= strlen(string_copy);
+  char user[USERNAME_LENGTH], host[HOSTNAME_LENGTH];
```
Should be `char user[USERNAME_LENGTH + 1], host[HOSTNAME_LENGTH + 1]`

laurynas-biveinis commented:
@gl-sergei, nice catch re. one char too-short buffers. Consider adding at least a single testcase with maximum length account and host name, if you haven't already

gl-sergei commented:
@laurynas-biveinis, already done :)
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1215/
